schema/crdb/dbinit.sql

```diff
@@ -2508,6 +2508,217 @@ CREATE TABLE IF NOT EXISTS omicron.public.switch_port_settings_address_config (
+
+/* Inventory Collections */
+
+-- list of all collections
+CREATE TABLE IF NOT EXISTS inv_collection (
+    id UUID PRIMARY KEY,
+    time_started TIMESTAMPTZ NOT NULL,
+    time_done TIMESTAMPTZ NOT NULL,
+    collector TEXT NOT NULL
+);
+-- Supports finding latest collection (to use) or the oldest collection (to
+-- clean up)
+CREATE INDEX IF NOT EXISTS inv_collection_by_time_started
+    ON omicron.public.inv_collection (time_started);
+
+-- list of errors generated during a collection
+CREATE TABLE IF NOT EXISTS omicron.public.inv_collection_error (
+    inv_collection_id UUID NOT NULL,
+    idx INT4 NOT NULL,
+    message TEXT
+);
+CREATE INDEX IF NOT EXISTS errors_by_collection
+    ON omicron.public.inv_collection_error (inv_collection_id, idx);
+
+/* what kind of slot MGS reported a device in */
+CREATE TYPE IF NOT EXISTS omicron.public.sp_type AS ENUM (
+    'sled',
+    'switch',
+    'power'
+);
+
+-- observations from and about service processors
+-- also see `inv_root_of_trust`
+CREATE TABLE IF NOT EXISTS omicron.public.inv_service_processor (
+    -- where this observation came from
+    -- (foreign key into `inv_collection` table)
+    inv_collection_id UUID NOT NULL,
+    -- which system this SP reports it is part of
+    -- (foreign key into `hw_baseboard_id` table)
+    hw_baseboard_id UUID NOT NULL,
+    -- when this observation was made
+    time_collected TIMESTAMPTZ NOT NULL,
+    -- which MGS instance reported this data
+    source TEXT NOT NULL,
+
+    -- identity of this device according to MGS
+    sp_type omicron.public.sp_type NOT NULL,
+    sp_slot INT4 NOT NULL,
+
+    -- Data from MGS "Get SP Info" API.  See MGS API documentation.
+    baseboard_revision INT8 NOT NULL,
+    hubris_archive_id TEXT NOT NULL,
+    power_state omicron.public.hw_power_state NOT NULL,
+
+    PRIMARY KEY (inv_collection_id, hw_baseboard_id)
+);
+
+-- root of trust information reported by SP
+-- There's usually one row here for each row in inv_service_processor, but not
+-- necessarily.
+CREATE TABLE IF NOT EXISTS omicron.public.inv_root_of_trust (
+    -- where this observation came from
+    -- (foreign key into `inv_collection` table)
+    inv_collection_id UUID NOT NULL,
+    -- which system this SP reports it is part of
+    -- (foreign key into `hw_baseboard_id` table)
+    hw_baseboard_id UUID NOT NULL,
+    -- when this observation was made
+    time_collected TIMESTAMPTZ NOT NULL,
+    -- which MGS instance reported this data
+    source TEXT NOT NULL,
+
+    slot_active omicron.public.hw_rot_slot NOT NULL,
+    slot_boot_pref_transient omicron.public.hw_rot_slot, -- nullable
+    slot_boot_pref_persistent omicron.public.hw_rot_slot NOT NULL,
+    slot_boot_pref_persistent_pending omicron.public.hw_rot_slot, -- nullable
+    slot_a_sha3_256 TEXT, -- nullable
+    slot_b_sha3_256 TEXT, -- nullable
+
+    PRIMARY KEY (inv_collection_id, hw_baseboard_id)
+);
+
+CREATE TYPE IF NOT EXISTS omicron.public.caboose_which AS ENUM (
```

Comment:
I went ahead and did this in 58c010f5f238f6ae5f17b93c16e045d0a4843aba.  Honestly, I could go either way on the result.  `CabooseWhich` does feel janky.  But it also reflects exactly what we're getting from MGS: that is, each row in this table reflects one response from the get-caboose endpoint, and that essentially represents the parameters to that request.  It's kind of a nice property that no row in an `inv_*` table represents data from multiple collection requests:

* it guides the schema design -- there's one table for each kind of observation (with maybe additional tables if there are a bunch of fields that can be present or absent together)
* it makes it easy to map the collection request responses to database rows
* it makes it easy to have the uniform set of (inv_collection_id, time_collected, source) fields.  We do have those here, but it's arguably misleading because the source and time_collected fields on `inv_service_processor` and `inv_root_of_trust` don't apply to the caboose fields
* This example sounds awfully specific but I feel like there's something general here: imagine if we wanted in the future to update the database _as_ we collect data instead of all at once at the end.  We'd have this unfortunate situation of having to either insert an `inv_service_processor` record and then update it later or else hang onto it (don't insert it) until we've tried to collected all the things that might go into it.

I don't think these are big deals for this particular case.  Rather, I came to this after exploring much different ways to structure this -- like an `inv_sled` table that might include pieces of information from both the sled agent (like the current host OS) and the SP (like the current host flash contents), etc.  I really disliked this because in the face of partial failures you have all these partial rows and then everything has to be NULL-able.  That's how I got to the "rows should not include data from multiple sources" rule.

Put differently: I don't think this specific violation of that rule is that bad, but without that rule, I found myself spinning in circles for a long time about how to design the schema.  It's pretty compelling to just say "each source of observation is a table; each observation is a row; then apply the usual database normalization rules".

All that said, I'm kind of ambivalent in the end.  I think I slightly prefer the previous thing with `caboose_which` but I'm interested in your thoughts.

Reply:
> I think it's the same either way. 

Yeah, from a representation perspective, that's true.  I had in my mind an implicit rule that we wouldn't want to write a record and then update it later in the same operation.  But that's somewhat arbitrary, too.  (_Not_ doing this does make it more complicated to infer anything from partially-inserted collections, and to measure progress based on what's present, but now we're talking about several layers of hypotheticals that aren't worth dealing with now.)